llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
       DL);
 }
 
+static SDValue
+PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
+  auto VT = N->getValueType(0);
+  if (!DCI.isAfterLegalizeDAG() || !Isv2x16VT(VT))
+    return SDValue();
+
+  auto Op0 = N->getOperand(0);
+  auto Op1 = N->getOperand(1);
+
+  // Start out by assuming we want to take the lower 2 bytes of each i32
+  // operand.
+  uint64_t Op0Bytes = 0x10;
+  uint64_t Op1Bytes = 0x54;
+
+  std::pair<SDValue *, uint64_t *> OpData[2] = {{&Op0, &Op0Bytes},
+                                                {&Op1, &Op1Bytes}};
+
+  // Check that each operand is an i16, truncated from an i32 operand. We'll
+  // select individual bytes from those original operands. Optionally, fold in a
+  // shift right of that original operand.
+  for (auto &[Op, OpBytes] : OpData) {
+    // Eat up any bitcast
+    if (Op->getOpcode() == ISD::BITCAST)
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
+        Op->getOperand(0).getValueType() != MVT::i32)
```

Comment:
In order to use it in a `prmt` we would need to extend it to `i32`. That is not a clear win. With truncate we reach out to the original `i32` we can use in permute.

With that in mind, we may not need to check for the single use here -- we're not adding any new register uses, we're just short-circuiting a part of the graph. If there are multiple uses for both 16-bit input values, there may be a slight benefit if we emit combining `mov`, but in most of the cases, `prmt` would allow us to skip some i32->i16 conversions and would be a net gain. I do not think we're at risk of increasing register pressure if we ignore one-use-ness of the inputs as prmt will be using the already existing `i32` values.

Reply:
> In order to use it in a prmt we would need to extend it to i32. That is not a clear win. With truncate we reach out to the original i32 we can use in permute.

Ahh, I see. Thank you for the explanation.

Reply:
Yep exactly. I was initially matching `i16` values but the generated PTX was always worse in that situation.